File: 131.分割回文串.py
#
# @lc app=leetcode.cn id=131 lang=python3
#
# [131] 分割回文串
#

# @lc code=start
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        ans = {tuple(s)}

        # initially separate string into chars
        # so that each char is always symmetric (list)
        # scan list vertify if 2 / 3 continuous element constitue symmetric string
        # if yes, merge these element and store into ans
        # if already exists in ans then no need to recursive 
        # since this circumstance has already been processed
        # if not exists recursive processed new list
        
        def isSymmetric(s1:str, s2:str):
            if s2[::-1] == s1:
                return True
            else:
                return False

        def recursive(loopList: list):
            if len(loopList) == 1:
                return None
            else:
                # next is even char
                for i in range(len(loopList) - 1):
                    if isSymmetric(loopList[i], loopList[i+1]):
                        loopListcopy = loopList[:]
                        loopListcopy[i] = loopList[i] + loopList[i+1]
                        del loopListcopy[i+1]
                        length = len(ans)
                        ans.add(tuple(loopListcopy))
                        if length != len(ans):
                            recursive(loopListcopy)
                    if i > 0:
                        if isSymmetric(loopList[i-1], loopList[i+1]):
                            loopListcopy = loopList[:]
                            loopListcopy[i] = loopList[i-1] + loopListcopy[i] + loopList[i+1]
                            del loopListcopy[i+1]
                            del loopListcopy[i-1]
                            length = len(ans)
                            ans.add(tuple(loopListcopy))
                            if length != len(ans):
                                recursive(loopListcopy)

        recursive(list(s))
        return [list(i) for i in ans]

class Solution:
    def isSymmtric(self, lo, hi):
        while lo < hi:
            if self.string[lo] != self.string[hi]:
                return False
            lo += 1
            hi -= 1
        return True

    # set a flag to divid string into processed and to_be_processed
    # scan to_be_processed
    # once reach a point that makes processed - point symmetric
    # recursive process remainder

    def recursive(self, start):
        if start == len(self.string):
            self.ans.append(self.stack[:])
            return None
        else:
            for i in range(start, len(self.string)):
                if self.isSymmtric(start, i):
                    self.stack.append(self.string[start:i+1])
                    self.recursive(i+1)
                    self.stack.pop()

    def partition(self, s: str) -> List[List[str]]:
        self.string = s
        self.ans = []
        self.stack = []
        self.recursive(0)
        return self.ans
# @lc code=end

File: 131.分割回文串.cpp
/*
 * @lc app=leetcode.cn id=131 lang=cpp
 *
 * [131] 分割回文串
 */

// @lc code=start
class Solution {
public:
    vector<vector<string>> partition(string s) {
        set<vector<string>> ansSet;
        vector<string> initial;
        for(auto chars:s){
            string charss(1, chars);
            initial.push_back(charss);
        }
        ansSet.insert(initial);
        this->recursive(initial, ansSet);
        return vector<vector<string>>(ansSet.begin(), ansSet.end());
    }

    void recursive(vector<string> initial, set<vector<string>>& ansSet){
        if(initial.size() == 1) return;
        for(int i = 0; i < initial.size()-1; ++i){
            if(isSymmetric(initial[i], initial[i+1])){
                vector<string> scopy = vector<string>(initial.begin(), initial.end());
                scopy[i] = scopy[i] + scopy[i+1];
                scopy.erase(scopy.begin()+i+1);
                int length = ansSet.size();
                ansSet.emplace(scopy);
                if(length != ansSet.size()){
                    recursive(scopy, ansSet);
                }
            }
            if(i > 0  && isSymmetric(initial[i-1], initial[i+1])){
                vector<string> scopy = vector<string>(initial.begin(), initial.end());
                scopy[i] = scopy[i-1] + scopy[i] + scopy[i+1];
                scopy.erase(scopy.begin()+i+1);
                scopy.erase(scopy.begin()+i-1);
                int length = ansSet.size();
                ansSet.emplace(scopy);
                if(length != ansSet.size()){
                    recursive(scopy, ansSet);
                } 
            }
        }
    }

    bool isSymmetric(string& s0, string& s1){
        if(s0.size() != s1.size()) return false;
        auto iteratorS0 = s0.begin();
        auto iteratorS1 = s1.rbegin();
        for(; iteratorS0 != s0.end(); iteratorS0++, iteratorS1++){
            if(*iteratorS0 != *iteratorS1) return false;
        }
        return true;
    }
};
// @lc code=end

